### PR TITLE
Add multi hot sparse categorical crossentropy

### PR DESCRIPTION
**Pull request by roywei**
### Summary
This PR adds a new feature to calculate categorical cross-entropy on multi hot sparse labels
Inputs are softmax predictions and true labels.
It should return same loss as categorical cross-entropy.

Example:
```
Input:
input data is random images of size (32, 32) in channels first data format
 Labels:
Tradition labels are in the shape of (num_samples, num_class), for example:
labels = [[0, 1, 1, ..., 0],
          [1, 1, 0, ..., 0],
          ...
          [0, 0, 0, ..., 1]]
where len(labels) = num_samples, len(labels[0]) = num_classes
 However, when num_classes are very large and labels are very sparse,
we can represent them differently, for example:
 There are total 1000 classes, so there will be 10000 different labels.
Each image can belong to at most 5 labels at the same time.
labels = [[1, 2],
          [0, 1],
          ...
          [999]]
where labels is a list of list
 Special Note:
To deal with different length of sparse labels, we pad them with negative values,
so we can differentiate padding values with normal labels. It will become:
padded_labels = pad_sequeences(labels, value=-1)
padded_labels = [[-1, -1, -1, 1, 2],
          [-1, -1, -1, 0, 1],
          ...
          [-1, -1, -1, -1, 999]]
It will have shape (num_samples, 5) which still save space compare to dense labels.
```
### Changes

- Add implementation for loss and metrics
- Add examples at examples/multi_hot_sparse_categorical_crossentropy.py
- Add unit tests for loss and metric
- Performance:
X3 faster for calculating 3000 classes multi labeled sparse labels. (5 labels at most for each data)
```
('categorical crossentropy loss time per epoch:', 0.7335219383239746)
('multi hot sparse categorical crossentropy loss time per epoch:', 0.23781204223632812)
```

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [x] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [x] This PR changes the current API [y/n]

### CODE REVIEW

File: keras/backend/mxnet_backend.py [hunk 1]
```diff
@@ -2987,6 +2987,47 @@ def sparse_categorical_crossentropy(target, output, from_logits=False, axis=-1):
...
+    # scale predictions so that the class probabilities of each sample sum to 1
+    if from_logits:
+        mx_output = mx.sym.softmax(mx_output, axis=axis)
+    else:
+        mx_output = mx.sym.broadcast_div(mx_output, mx.sym.sum(mx_output,
+                                                               axis=axis,
+                                                               keepdims=True))
+    mx_output = mx.sym.clip(mx_output, a_min=epsilon(), a_max=1.0 - epsilon())
+    mx_output = mx.sym.concat(mx.sym.full((target.shape[0],1), 0.5), mx_output)
+    from mxnet.symbol.contrib import foreach
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: nit: move it.

File: keras/backend/mxnet_backend.py [hunk 2]
```diff
@@ -2987,6 +2987,47 @@ def sparse_categorical_crossentropy(target, output, from_logits=False, axis=-1):
...
+
+    mx_output = output.symbol
+    # scale predictions so that the class probabilities of each sample sum to 1
+    if from_logits:
+        mx_output = mx.sym.softmax(mx_output, axis=axis)
+    else:
+        mx_output = mx.sym.broadcast_div(mx_output, mx.sym.sum(mx_output,
+                                                               axis=axis,
+                                                               keepdims=True))
+    mx_output = mx.sym.clip(mx_output, a_min=epsilon(), a_max=1.0 - epsilon())
```
Comment (sandeep-krishnamurthy) [inline keras/backend/mxnet_backend.py]: Nice work. Please document these steps so it will be easier later for maintenance.

File: examples/multi_hot_sparse_categorical_crossentropy.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    # each data have at most 5 labels
+    for j in range(0, 5):
+        label = random.randint(0, num_classes - 1)
+        dense_labels[i][label] = 1
+        # making the number of labels for each data unequal
+        if random.randint(0, 5) == 1:
+            break
+    sparse_label_j = np.where(dense_labels[i] == 1)[0]
+    sparse_labels.append(sparse_label_j)
+
```
Comment (kalyc) [inline examples/multi_hot_sparse_categorical_crossentropy.py]: [minor] remove blank line

File: examples/multi_hot_sparse_categorical_crossentropy.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+          epochs=epoch)
+
+
+# use normal multi_hot_sparse_categorical_crossentropy
+model.compile(loss=keras.losses.multi_hot_sparse_categorical_crossentropy,
+              optimizer=keras.optimizers.Adadelta(),
+              metrics=['accuracy'])
+
+# pad sparse labels into shape length with value -1 to differentiate from normal labels
+y_train_pad = pad_sequences(sparse_labels, value=-1)
```
Comment (kalyc) [inline examples/multi_hot_sparse_categorical_crossentropy.py]: why not add this padding logic inside multi_hot_sparse_categorical_crossentropy calculating method?

File: keras/backend/mxnet_backend.py [hunk 5]
```diff
@@ -2988,6 +2988,66 @@ def sparse_categorical_crossentropy(target, output, from_logits=False, axis=-1):
...
+        mx_output = mx.sym.broadcast_div(mx_output, mx.sym.sum(mx_output,
+                                                               axis=axis,
+                                                               keepdims=True))
+    # clip to prevent NaN's and Inf's
+    mx_output = mx.sym.clip(mx_output, a_min=epsilon(), a_max=1.0 - epsilon())
+
+    # using control flow ops to iterate output and take target (true label)
+    _step = lambda data, _: (mx.sym.take(data[0], data[1]), [])
+    data = [mx_output, target.symbol]
+    outputs, _ = mx.symbol.contrib.foreach(_step, data, [])
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: this is a MXNet contrib API do we have an issue to track that this needs to be updated when the stable version is released for foreach?

File: tests/keras/losses_test.py [hunk 6]
```diff
@@ -109,6 +109,22 @@ def test_sparse_categorical_crossentropy_4d():
     assert np.isclose(expected_loss, np.mean(loss))
 
 
+def test_multi_hot_sparse_categorical_crossentropy():
```
Comment (kalyc) [inline tests/keras/losses_test.py]: please add test for multi_hot_sparse_categorical_accuracy

File: keras/backend/mxnet_backend.py (hunk 5 above)
Comment (roywei) [reply keras/backend/mxnet_backend.py]: Currently we have CI to test MXNet API changes (deprecated or breaking change). It's the same for contrib ops and other normal ops. If it's moved to mx.sym, it will give a deprecated warning. we should be able to catch it. created https://github.com/awslabs/keras-apache-mxnet/issues/173

File: tests/keras/losses_test.py (hunk 6 above)
Comment (roywei) [reply tests/keras/losses_test.py]: added, see changes at tests/keras/metrics_test.py

File: examples/multi_hot_sparse_categorical_crossentropy.py (hunk 3 above)
Comment (roywei) [reply examples/multi_hot_sparse_categorical_crossentropy.py]: removed

File: keras/backend/mxnet_backend.py [hunk 7]
```diff
@@ -3017,6 +3017,11 @@ def multi_hot_sparse_categorical_crossentropy(target, output, from_logits=False,
     >>>y_true_np2 = np.array([[1, 2], [0, 2],[0]])
     ```
     """
+    # TODO: remove version check after mxnet 1.3.1 stable release
```
Comment (kalyc) [inline keras/backend/mxnet_backend.py]: please create a tracking issue to remove this check

Comment (roywei) [reply keras/backend/mxnet_backend.py]: tracked at https://github.com/awslabs/keras-apache-mxnet/issues/175
